sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableSharedClientTest.java

```diff
@@ -0,0 +1,264 @@
+import org.apache.beam.runners.direct.DirectRunner;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.PipelineResult.State;
+import org.apache.beam.sdk.io.GenerateSequence;
+import org.apache.beam.sdk.options.ExperimentalOptions;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableList;
+import org.hamcrest.Matchers;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Ensure that BigtableIO.write() reuses the same instance of the underlying bigtable client. This
+ * test will create a toy pipeline using DirectRunner and have it write to a local emulator. The
+ * emulator will record all of the client connections. Then the test will check that only a single
+ * connection was used.
+ */
+@RunWith(JUnit4.class)
+public class BigtableSharedClientTest {
+
+  private FakeBigtable fakeService;
+  private ServerClientConnectionCounterInterceptor clientConnectionInterceptor;
+  private Server fakeServer;
+
+  @Before
+  public void setUp() throws Exception {
+    clientConnectionInterceptor = new ServerClientConnectionCounterInterceptor();
+    this.fakeService = new FakeBigtable();
+
+    IOException lastError = null;
+
+    for (int i = 0; i < 10; i++) {
+      try {
+        this.fakeServer = createServer(fakeService, clientConnectionInterceptor);
+        lastError = null;
+        break;
+      } catch (IOException e) {
+        lastError = e;
+      }
+    }
+    if (lastError != null) {
+      throw lastError;
+    }
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    if (fakeServer != null) {
+      fakeServer.shutdownNow();
+    }
+  }
+
+  private static Server createServer(BindableService service, ServerInterceptor... interceptors)
+      throws IOException {
+    int port;
+    try (ServerSocket ss = new ServerSocket(0)) {
+      port = ss.getLocalPort();
+    }
+
+    ServerBuilder<?> serverBuilder = ServerBuilder.forPort(port).addService(service);
+
+    for (ServerInterceptor interceptor : interceptors) {
+      serverBuilder.intercept(interceptor);
+    }
+    return serverBuilder.build().start();
+  }
+
+  @Test
+  public void testClientReusedAcrossBundles() {
+    PipelineOptions opts = PipelineOptionsFactory.create();
+    opts.setRunner(DirectRunner.class);
+    ExperimentalOptions.addExperiment(
+        opts.as(ExperimentalOptions.class),
+        String.format(
+            "%s=%s",
+            BigtableConfigTranslator.BIGTABLE_SETTINGS_OVERRIDE,
+            ClientSettingsOverride.class.getName()));
+
+    Pipeline pipeline = Pipeline.create(opts);
+
+    MutationsDoFn dofn = new MutationsDoFn();
```

Comment:
this could just be inline in the ParDo.of?

Reply:
that gets a little messy because I need a static counter, I think its cleaner to keep it as an inner class that owns its own static counter

Reply:
oh I just mean the `ParDo.of(new MutationsDoFn())`

Reply:
oh ok, done